From 0f45013096dee23f9172f6a8b0d9736bb1e736c7 Mon Sep 17 00:00:00 2001
From: Henry Wu <Henry_Wu@quantatw.com>
Date: Mon, 26 Jan 2026 15:47:59 +0800
Subject: [PATCH 02/19] oem-meta: santabarbara: add handler for event logs from
 rainbow MMC

Add support for decoding the event logs emitted by the Rainbow MMC. To
distinguish between MITA and Rainbow event formats, the function
lookupFaultByIndex is updated so that it can retrieve the corresponding
event type and provide the correct interpretation.

This commit is base on
82142: oem-meta: Add file IO responder for event logs from BIC
https://gerrit.openbmc.org/c/openbmc/pldm/+/82142

Tested by triggering MITA fault logs to verify correct event type
handling.
```
    "7": {
        "additional_data": {
            "FAILURE_DATA": "MTIA_PVDD1P5_S - [02 03]",
            "VOLTAGE_REGULATOR": "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pvdd1p5_s",
            "_CODE_FILE": "../../../../../../workspace/sources/pldm/oem/meta/libpldmresponder/file_io_type_event_log.cpp",
            "_CODE_FUNC": "void pldm::responder::oem_meta::recordEventLog(const FaultDesc&, EventAssert, const FaultData&)",
            "_CODE_LINE": "451",
            "_PID": "1267"
        },
        "event_id": "",
        "message": "xyz.openbmc_project.State.Power.VoltageRegulatorFault",
        "redfish": {
            "args": [
                "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pvdd1p5_s"
            ],
            "id": "OpenBMC_StatePower.VoltageRegulatorFault",
            "message": "A fault has been detected on voltage regulator '/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pvdd1p5_s'"
        },
        "resolution": "",
        "resolved": false,
        "severity": "xyz.openbmc_project.Logging.Entry.Level.Error",
        "timestamp": "2026-01-26T05:38:09.990000000Z",
        "updated_timestamp": "2026-01-26T05:38:09.990000000Z"
    }
}
```

Tested by triggering Rainbow fault logs to verify correct event
type handling.
```
"86": {
  "additional_data": {
    "FAILURE_DATA": "RAINBOW_OWL_E_TRVDD0P75 - [7C 00]",
    "VOLTAGE_REGULATOR": "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow_0/owl_e_trvdd0p75",
    "_CODE_FILE": "../../../../../../workspace/sources/pldm/oem/meta/libpldmresponder/file_io_type_event_log.cpp",
    "_CODE_FUNC": "void pldm::responder::oem_meta::recordEventLog(const FaultDesc&, EventAssert, const FaultData&)",
    "_CODE_LINE": "439",
    "_PID": "1038"
  },
  "event_id": "",
  "message": "xyz.openbmc_project.State.Power.VoltageRegulatorFault",
  "redfish": {
    "args": [
      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow_0/owl_e_trvdd0p75"
    ],
    "id": "OpenBMC_StatePower.VoltageRegulatorFault",
    "message": "A fault has been detected on voltage regulator '/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow_0/owl_e_trvdd0p75'"
  },
  "resolution": "",
  "resolved": false,
  "severity": "xyz.openbmc_project.Logging.Entry.Level.Error",
  "timestamp": "2025-11-26T01:34:28.595000000Z",
  "updated_timestamp": "2025-11-26T01:34:28.595000000Z"
}
```

Change-Id: Ie5d6e6fe26baf1cb7c23d19127d0d84f66f37c27
Signed-off-by: Henry Wu <Henry_Wu@quantatw.com>
Signed-off-by: Patrick Williams <patrick@stwcx.xyz>
---
 .../file_io_type_event_log.cpp                | 226 ++++++++++++++++++
 1 file changed, 226 insertions(+)

diff --git a/oem/meta/libpldmresponder/file_io_type_event_log.cpp b/oem/meta/libpldmresponder/file_io_type_event_log.cpp
index d0ebe95..9414edd 100644
--- a/oem/meta/libpldmresponder/file_io_type_event_log.cpp
+++ b/oem/meta/libpldmresponder/file_io_type_event_log.cpp
@@ -64,6 +64,7 @@ enum class EventType
     OS_LOAD_WDT_PWR_CYCLE = 0x26,
     MTIA_FAULT = 0x27,
     POST_TIMEOUTED = 0x28,
+    RAINBOW_FAULT = 0x29,
 };
 
 enum class EventAssert : uint8_t
@@ -120,6 +121,7 @@ static constexpr auto eventList = std::to_array({
     "OS LOAD WDT PWR_CYCLE",
     "MTIA FAULT",
     "Post-Timeouted",
+    "RAINBOW FAULT",
 });
 
 static constexpr auto pmic_event_type = std::to_array({
@@ -424,6 +426,185 @@ static const std::map<uint8_t, FaultDesc> mtiaFaultsList = {
       "DEVICE", record::thermalFault}},
 };
 
+static const std::map<uint8_t, FaultDesc> rainbowFaultsList = {
+    {0x00,
+     {"RAINBOW_HAMSA_VDDHRXTX_PCIE",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/hamsa_vddhrxtx_pcie",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x01,
+     {"RAINBOW_HAMSA_AVDD_PCIE",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/hamsa_avdd_pcie",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x02,
+     {"RAINBOW_OWL_W_TRVDD0P75",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/owl_w_trvdd0p75",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x03,
+     {"RAINBOW_OWL_E_TRVDD0P75",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/owl_e_trvdd0p75",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x04,
+     {"RAINBOW_OWL_W_TRVDD0P9",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/owl_w_trvdd0p9",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x05,
+     {"RAINBOW_OWL_E_TRVDD0P9",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/owl_e_trvdd0p9",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x06,
+     {"RAINBOW_MAX_N_VDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/max_n_vdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x07,
+     {"RAINBOW_MAX_M_VDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/max_m_vdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x08,
+     {"RAINBOW_MAX_S_VDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/max_s_vdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x09,
+     {"RAINBOW_HAMSA_VDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/hamsa_vdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0A,
+     {"RAINBOW_OWL_W_VDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/owl_w_vdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0B,
+     {"RAINBOW_OWL_E_VDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/owl_e_vdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0C,
+     {"RAINBOW_MEDHA0_VDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/medha0_vdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0D,
+     {"RAINBOW_MEDHA1_VDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/medha1_vdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0E,
+     {"RAINBOW_VDDPHY_HBM1_HBM3_HBM5_HBM7",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/vddphy_hbm1_hbm3_hbm5_hbm7",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0F,
+     {"RAINBOW_VPP_HBM1_HBM3_HBM5_HBM7",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/vpp_hbm1_hbm3_hbm5_hbm7",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x10,
+     {"RAINBOW_VDDQC_HBM1_HBM3_HBM5_HBM7",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/vddqc_hbm1_hbm3_hbm5_hbm7",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x11,
+     {"RAINBOW_VDDQL_HBM1_HBM3_HBM5_HBM7",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/vddql_hbm1_hbm3_hbm5_hbm7",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x12,
+     {"RAINBOW_VDDPHY_HBM0_HBM2_HBM4_HBM6",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/vddphy_hbm0_hbm2_hbm4_hbm6",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x13,
+     {"RAINBOW_VPP_HBM0_HBM2_HBM4_HBM6",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/vpp_hbm0_hbm2_hbm4_hbm6",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x14,
+     {"RAINBOW_VDDQC_HBM0_HBM2_HBM4_HBM6",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/vddqc_hbm0_hbm2_hbm4_hbm6",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x15,
+     {"RAINBOW_VDDQL_HBM0_HBM2_HBM4_HBM6",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/vddql_hbm0_hbm2_hbm4_hbm6",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x16,
+     {"RAINBOW_P1V5_W_RVDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p1v5_w_rvdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x17,
+     {"RAINBOW_P1V5_E_RVDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p1v5_e_rvdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x18,
+     {"RAINBOW_P0V9_OWL_W_PVDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p0v9_owl_w_pvdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x19,
+     {"RAINBOW_P0V9_OWL_E_PVDD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p0v9_owl_e_pvdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x1A,
+     {"RAINBOW_PLL_VDDA15_HBM5_HBM7",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/pll_vdda15_hbm5_hbm7",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x1B,
+     {"RAINBOW_PLL_VDDA15_HBM1_HBM3",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/pll_vdda15_hbm1_hbm3",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x1C,
+     {"RAINBOW_PLL_VDDA15_HBM4_HBM6",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/pll_vdda15_hbm4_hbm6",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x1D,
+     {"RAINBOW_PLL_VDDA15_HBM0_HBM2",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/pll_vdda15_hbm0_hbm2",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x1E,
+     {"RAINBOW_PVDD1P5",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/pvdd1p5",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x1F,
+     {"RAINBOW_P1V5_PLL_VDDA_SOC",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p1v5_pll_vdda_soc",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x20,
+     {"RAINBOW_P1V5_PLL_VDDA_OWL",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p1v5_pll_vdda_owl",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x21,
+     {"RAINBOW_LDO_IN_1V2",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/ldo_in_1v2",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x22,
+     {"RAINBOW_P1V8",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p1v8",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x23,
+     {"RAINBOW_P3V3",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p3v3",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x24,
+     {"RAINBOW_P5V",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p5v",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x25,
+     {"RAINBOW_P12V_UBC_PWRGD",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p12v_ubc_pwrgd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x26,
+     {"RAINBOW_P0V75_AVDD_HCSL ",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p0v75_avdd_hcsl",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x27,
+     {"RAINBOW_P4V2",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/p4v2",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x50,
+     {"RAINBOW_POWER_ON_SEQUENCE_FAIL",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/power_on_sequence_fail",
+      "POWER_RAIL", record::powerRailFault}},
+    {0x51,
+     {"RAINBOW_FM_ASIC_0_THERMTRIP_N",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/fm_asic_0_thermtrip_n",
+      "DEVICE", record::thermalFault}},
+    {0x52,
+     {"RAINBOW_MEDHA1_HBM_CATTRIP_LS_LVC33_ALARM",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/medha1_hbm_cattrip_ls_lvc33_alarm",
+      "DEVICE", record::thermalFault}},
+    {0x53,
+     {"RAINBOW_MEDHA0_HBM_CATTRIP_LS_LVC33_ALARM",
+      "/xyz/openbmc_project/inventory/system/board/Santabarbara_Rainbow/medha0_hbm_cattrip_ls_lvc33_alarm",
+      "DEVICE", record::thermalFault}},
+};
+
 namespace report
 {
 static void recordEventLog(const FaultDesc& fault, EventAssert eventStatus,
@@ -499,6 +680,50 @@ static void mtiaFault(const message& eventData, const std::string&,
     recordEventLog(faultEvent->second, eventStatus, eventData[3], eventData[4]);
 }
 
+static void rainbowFault(const message& eventData, const std::string&,
+                         const std::string& slotNumStr)
+{
+    EventAssert eventStatus = static_cast<EventAssert>(eventData[1]);
+    auto faultEvent = rainbowFaultsList.find(eventData[2]);
+    if (faultEvent == std::end(rainbowFaultsList))
+    {
+        lg2::error("Invalid Rainbow fault source 0x{SOURCE}", "SOURCE",
+                   lg2::hex, eventData[2]);
+        return;
+    }
+
+    int slot = -1;
+    try
+    {
+        slot = std::stoi(slotNumStr);
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error("Invalid slot number {SLOT}", "SLOT", slotNumStr);
+    }
+
+    FaultDesc updatedFault = faultEvent->second;
+
+    if (slot != -1)
+    {
+        constexpr std::string_view needle = "/Santabarbara_Rainbow";
+        std::string newPath = updatedFault.object_path;
+        size_t pos = newPath.find(needle);
+
+        if (pos != std::string::npos)
+        {
+            newPath.insert(pos + needle.size(), "_" + std::to_string(slot));
+        }
+        else
+        {
+            newPath += "_" + std::to_string(slot);
+        }
+        updatedFault.object_path = std::move(newPath);
+    }
+
+    recordEventLog(updatedFault, eventStatus, eventData[3], eventData[4]);
+}
+
 static void reportError(const message&, const std::string& event,
                         const std::string&)
 {
@@ -569,6 +794,7 @@ static const std::map<EventType, EventDescriptor> eventHandlers = {
     {EventType::PROCHOT_IS_TRIGGERED_DUE_TO_SD_SENSOR_READING_EXCEED_UCR,
      {report::prochotSdSensor, nullptr}},
     {EventType::MTIA_FAULT, {report::mtiaFault, nullptr}},
+    {EventType::RAINBOW_FAULT, {report::rainbowFault, nullptr}},
     {EventType::PMALERT_ASSERTION, {report::reportError, format::vrFault}},
     {EventType::FRB3_TIMER_EXPIRED, {report::reportError, nullptr}},
     {EventType::ADDC_DUMP, {report::reportError, nullptr}},
-- 
2.52.0

